File: pettime-backend/src/utils/validators.js
// Validadores de dados do sistema
function validarEmail(email) {
  if (!email || typeof email !== 'string') {
    return false;
  }
  
  const emailTrimmed = email.trim();
  if (emailTrimmed === '') {
    return false;
  }
  
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  
  // Verificações adicionais
  if (emailTrimmed.includes('..') || 
      emailTrimmed.startsWith('.') || 
      emailTrimmed.endsWith('.') ||
      emailTrimmed.startsWith('@') ||
      emailTrimmed.endsWith('@')) {
    return false;
  }
  
  return emailRegex.test(emailTrimmed);
}

function validarTelefone(telefone) {
  if (!telefone || typeof telefone !== 'string') {
    return false;
  }
  
  // Remove caracteres não numéricos
  const numeroLimpo = telefone.replace(/\D/g, '');
  
  // Verifica se tem 10 ou 11 dígitos
  if (numeroLimpo.length !== 10 && numeroLimpo.length !== 11) {
    return false;
  }
  
  // Verifica DDD (códigos de área válidos do Brasil)
  const ddd = parseInt(numeroLimpo.substring(0, 2));
  const dddsValidos = [
    11, 12, 13, 14, 15, 16, 17, 18, 19, // SP
    21, 22, 24, // RJ/ES
    27, 28, // ES
    31, 32, 33, 34, 35, 37, 38, // MG
    41, 42, 43, 44, 45, 46, // PR
    47, 48, 49, // SC
    51, 53, 54, 55, // RS
    61, // DF/GO
    62, 64, // GO/TO
    63, // TO
    65, 66, // MT
    67, // MS
    68, // AC
    69, // RO
    71, 73, 74, 75, 77, // BA
    79, // SE
    81, 87, // PE
    82, // AL
    83, // PB
    84, // RN
    85, 88, // CE
    86, 89, // PI
    91, 93, 94, // PA
    92, 97, // AM
    95, // RR
    96, // AP
    98, 99 // MA
  ];
  
  if (!dddsValidos.includes(ddd)) {
    return false;
  }
  
  // Para celular (11 dígitos), o terceiro dígito deve ser 9
  if (numeroLimpo.length === 11) {
    if (numeroLimpo[2] !== '9') {
      return false;
    }
  }
  
  // Para telefone fixo (10 dígitos), verificar se não começa com 9
  if (numeroLimpo.length === 10) {
    if (numeroLimpo[2] === '9') {
      return false;
    }
  }
  
  return true;
}

function validarCPF(cpf) {
  if (!cpf || typeof cpf !== 'string') {
    return false;
  }
  
  // Remove caracteres não numéricos
  const cpfLimpo = cpf.replace(/\D/g, '');
  
  // Verifica se tem 11 dígitos
  if (cpfLimpo.length !== 11) {
    return false;
  }
  
  // Verifica se todos os dígitos são iguais
  if (/^(\d)\1{10}$/.test(cpfLimpo)) {
    return false;
  }
  
  // Calcula o primeiro dígito verificador
  let soma = 0;
  for (let i = 0; i < 9; i++) {
    soma += parseInt(cpfLimpo[i]) * (10 - i);
  }
  let resto = (soma * 10) % 11;
  if (resto === 10 || resto === 11) resto = 0;
  if (resto !== parseInt(cpfLimpo[9])) return false;
  
  // Calcula o segundo dígito verificador
  soma = 0;
  for (let i = 0; i < 10; i++) {
    soma += parseInt(cpfLimpo[i]) * (11 - i);
  }
  resto = (soma * 10) % 11;
  if (resto === 10 || resto === 11) resto = 0;
  if (resto !== parseInt(cpfLimpo[10])) return false;
  
  return true;
}

function validarSenha(senha) {
  if (!senha || typeof senha !== 'string') {
    return false;
  }
  
  // Mínimo 8 caracteres
  if (senha.length < 8) {
    return false;
  }
  
  // Máximo 50 caracteres
  if (senha.length > 50) {
    return false;
  }
  
  // Deve conter pelo menos uma letra
  if (!/[a-zA-Z]/.test(senha)) {
    return false;
  }
  
  // Deve conter pelo menos um número
  if (!/[0-9]/.test(senha)) {
    return false;
  }
  
  return true;
}

function validarNome(nome) {
  if (!nome || typeof nome !== 'string') {
    return false;
  }
  
  const nomeTrimmed = nome.trim();
  
  // Mínimo 2 caracteres
  if (nomeTrimmed.length < 2) {
    return false;
  }
  
  // Máximo 100 caracteres
  if (nomeTrimmed.length > 100) {
    return false;
  }
  
  // Deve conter pelo menos um sobrenome
  if (!nomeTrimmed.includes(' ')) {
    return false;
  }
  
  // Só letras, espaços e hífens
  if (!/^[a-zA-ZÀ-ÿ\s\-]+$/.test(nomeTrimmed)) {
    return false;
  }
  
  return true;
}

module.exports = {
  validarEmail,
  validarTelefone,
  validarCPF,
  validarSenha,
  validarNome
};

File: pettime_frontend/lib/core/utils/validators.dart
class FormValidator {
  /// Valida se uma string não está vazia
  static String? validarCampoObrigatorio(String? valor, [String? nomeCampo]) {
    if (valor == null || valor.trim().isEmpty) {
      return '${nomeCampo ?? 'Este campo'} é obrigatório';
    }
    return null;
  }

  /// Valida formato de email
  static String? validarEmail(String? email) {
    if (email == null || email.isEmpty) {
      return 'Email é obrigatório';
    }

    final emailRegex = RegExp(
      r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$',
    );

    if (!emailRegex.hasMatch(email)) {
      return 'Digite um email válido';
    }

    // Verifica comprimento máximo
    if (email.length > 254) {
      return 'Email muito longo';
    }

    return null;
  }

  /// Valida senha com critérios de segurança rigorosos
  static String? validarSenha(String? senha) {
    if (senha == null || senha.isEmpty) {
      return 'Senha é obrigatória';
    }

    if (senha.length < 8) {
      return 'A senha deve ter no mínimo 8 caracteres';
    }

    if (senha.length > 50) {
      return 'Senha muito longa (máximo 50 caracteres)';
    }

    // Verifica se tem pelo menos uma letra maiúscula
    if (!RegExp(r'[A-Z]').hasMatch(senha)) {
      return 'Senha deve conter pelo menos uma letra maiúscula';
    }

    // Verifica se tem pelo menos uma letra minúscula
    if (!RegExp(r'[a-z]').hasMatch(senha)) {
      return 'Senha deve conter pelo menos uma letra minúscula';
    }

    // Verifica se tem pelo menos um número
    if (!RegExp(r'\d').hasMatch(senha)) {
      return 'Senha deve conter pelo menos um número';
    }

    // Verifica se tem pelo menos um símbolo
    if (!RegExp(
      r'''[!@#\$%\^&\*\(\)_\+\-=\[\]{};:'",.<>?]''',
    ).hasMatch(senha)) {
      return 'Senha deve conter pelo menos um símbolo';
    }

    return null;
  }

  /// Valida confirmação de senha
  static String? validarConfirmacaoSenha(String? senha, String? confirmacao) {
    if (confirmacao == null || confirmacao.isEmpty) {
      return 'Confirmação de senha é obrigatória';
    }

    if (senha != confirmacao) {
      return 'Senhas não coincidem';
    }

    return null;
  }

  /// Valida telefone brasileiro
  static String? validarTelefone(String? telefone) {
    if (telefone == null || telefone.isEmpty) {
      return 'Telefone é obrigatório';
    }

    // Remove caracteres especiais
    final numeros = telefone.replaceAll(RegExp(r'[^\d]'), '');

    if (numeros.length < 10 || numeros.length > 11) {
      return 'Telefone deve ter 10 ou 11 dígitos';
    }

    // Valida código de área (11-99)
    final codigoArea = int.tryParse(numeros.substring(0, 2));
    if (codigoArea == null || codigoArea < 11 || codigoArea > 99) {
      return 'Código de área inválido';
    }

    // Para celular (11 dígitos), o primeiro dígito após o código deve ser 9
    if (numeros.length == 11) {
      if (numeros[2] != '9') {
        return 'Número de celular inválido';
      }
    }

    return null;
  }

  /// Valida nome completo
  static String? validarNome(String? nome) {
    if (nome == null || nome.trim().isEmpty) {
      return 'Nome é obrigatório';
    }

    final nomeLimpo = nome.trim();

    if (nomeLimpo.length < 2) {
      return 'Nome deve ter pelo menos 2 caracteres';
    }

    if (nomeLimpo.length > 100) {
      return 'Nome muito longo (máximo 100 caracteres)';
    }

    // Verifica se contém apenas letras (incluindo acentos), espaços e hífens
    if (!RegExp(r'^[a-zA-ZÀ-ÿ\s\-\.]+$').hasMatch(nomeLimpo)) {
      return 'Nome contém caracteres inválidos';
    }

    // Verifica se tem pelo menos um espaço (nome e sobrenome)
    final palavras = nomeLimpo.split(' ').where((p) => p.isNotEmpty).toList();
    if (palavras.length < 2) {
      return 'Digite nome e sobrenome';
    }

    return null;
  }

  /// Valida data de nascimento
  static String? validarDataNascimento(DateTime? data) {
    if (data == null) {
      return 'Data de nascimento é obrigatória';
    }

    final hoje = DateTime.now();
    final idade = hoje.year - data.year;

    // Verifica se não é no futuro
    if (data.isAfter(hoje)) {
      return 'Data não pode ser no futuro';
    }

    // Verifica idade mínima (0 anos) e máxima (150 anos)
    if (idade < 0) {
      return 'Data inválida';
    }

    if (idade > 150) {
      return 'Data muito antiga';
    }

    return null;
  }

  /// Valida peso em quilos
  static String? validarPeso(String? pesoStr) {
    if (pesoStr == null || pesoStr.trim().isEmpty) {
      return null; // Peso é opcional
    }

    final peso = double.tryParse(pesoStr.replaceAll(',', '.'));

    if (peso == null) {
      return 'Digite um peso válido';
    }

    if (peso <= 0) {
      return 'Peso deve ser maior que zero';
    }

    if (peso > 1000) {
      return 'Peso muito alto (máximo 1000kg)';
    }

    return null;
  }

  /// Valida CPF brasileiro
  static String? validarCPF(String? cpf) {
    if (cpf == null || cpf.isEmpty) {
      return 'CPF é obrigatório';
    }

    // Remove caracteres especiais
    final numeros = cpf.replaceAll(RegExp(r'[^\d]'), '');

    if (numeros.length != 11) {
      return 'CPF deve ter 11 dígitos';
    }

    // Verifica se todos os dígitos são iguais
    if (RegExp(r'^(\d)\1*$').hasMatch(numeros)) {
      return 'CPF inválido';
    }

    // Validação dos dígitos verificadores
    if (!_validarDigitosCPF(numeros)) {
      return 'CPF inválido';
    }

    return null;
  }

  static bool _validarDigitosCPF(String cpf) {
    // Calcula primeiro dígito verificador
    int soma = 0;
    for (int i = 0; i < 9; i++) {
      soma += int.parse(cpf[i]) * (10 - i);
    }
    int primeiroDigito = (soma * 10) % 11;
    if (primeiroDigito == 10) primeiroDigito = 0;

    if (primeiroDigito != int.parse(cpf[9])) return false;

    // Calcula segundo dígito verificador
    soma = 0;
    for (int i = 0; i < 10; i++) {
      soma += int.parse(cpf[i]) * (11 - i);
    }
    int segundoDigito = (soma * 10) % 11;
    if (segundoDigito == 10) segundoDigito = 0;

    return segundoDigito == int.parse(cpf[10]);
  }

  /// Valida CEP brasileiro
  static String? validarCEP(String? cep) {
    if (cep == null || cep.isEmpty) {
      return null; // CEP é opcional
    }

    // Remove caracteres especiais
    final numeros = cep.replaceAll(RegExp(r'[^\d]'), '');

    if (numeros.length != 8) {
      return 'CEP deve ter 8 dígitos';
    }

    // Verifica se não são todos zeros
    if (numeros == '00000000') {
      return 'CEP inválido';
    }

    return null;
  }

  /// Valida horário no formato HH:mm
  static String? validarHorario(String? horario) {
    if (horario == null || horario.isEmpty) {
      return 'Horário é obrigatório';
    }

    final horarioRegex = RegExp(r'^([01]\d|2[0-3]):([0-5]\d)$');

    if (!horarioRegex.hasMatch(horario)) {
      return 'Digite um horário válido (HH:mm)';
    }

    return null;
  }
}
